I18n Support for Angular Authentication Components

seang96 opened the pull request:
Description of changes:
Fixed build errors when trying to build on Windows (needed to do this for local development)
Added I18n inside provider AmplifyService
Added I18n support to aws-amplify-angular authentication components
Removed error message being an alert for ionic components
Made error message boxes the same throughout authentication components and in the correct location inside the template. Some were inside the amplify-form-body.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

sprucify commented:
Thanks so much for creating this feature. I18n is essential!

sprucify commented on packages/aws-amplify-angular/src/assets/default-sign-up-fields.ts:
```diff
@@ -12,30 +12,31 @@
  * and limitations under the License.
  */
 // tslint:enable
+import { I18n } from '@angular/core';
```
Does @angular/core have a I18n module?

seang96 replied:
@sprucify  @angular/core does have an existing I18n module. This has already been done to the ract/vue components and has a docs page! [https://aws-amplify.github.io/docs/js/i18n](https://aws-amplify.github.io/docs/js/i18n)

sprucify replied:
Yes, but not from package @angular/core?

seang96 replied:
@sprucify Oh I did misread the import at first and put it in @angular/core, however I changed the imports to the correct ones afterwards. See the final commit of the file: [https://github.com/aws-amplify/amplify-js/blob/fdfd2122b5b14d2eb08d52ab31fc8caf9fbab935/packages/aws-amplify-angular/src/assets/default-sign-up-fields.ts](https://github.com/aws-amplify/amplify-js/blob/fdfd2122b5b14d2eb08d52ab31fc8caf9fbab935/packages/aws-amplify-angular/src/assets/default-sign-up-fields.ts)

sprucify replied:
Ah sorry, I overlooked your more recent commit.
I’m going to implement this in my app and hopefully it will be part of master soon

sprucify replied:
I couldn’t get it working by replacing the version numbering with your repository url. (Inexperience me)
I guess I have to wait for the amplify update.

seang96 replied:
Put this in your package.json script section, npm install to install first time, updates to branch use npm run update-amplify-js.
```
    "preinstall": "git clone https://github.com/seang96/amplify-js.git --single-branch --branch angular-auth-I18n && npm run build-amplify-js",
    "update-amplify-js": "cd amplify-js && git fetch && git reset --hard origin/angular-auth-I18n && git clean -df && cd .. && npm run build-amplify-js",
    "build-amplify-js": "cd amplify-js && yarn config set workspaces-experimental true && yarn && npx yarn bootstrap && yarn build",
```

haverchuck commented:
@seang96  - Thanks so much for your PR.  This is definitely a needed addition.  It is being reviewed now - comments will be left in line.

haverchuck commented on packages/aws-amplify-angular/src/assets/default-sign-up-fields.ts:
```diff
@@ -12,30 +12,33 @@
   },
   {
-    label: 'Email',
+    label: 'Email *',
```
@seang96  -  Please revert the change to these label values.  They should not have an asterix.  The required attribute should conditionally insert an asterix.  (This will be another change to remove.)

seang96 replied:
@haverchuck I explicitly defined these with an asterisk if they were required rather than the span element containing the element in case another language didn't use asterisks for field requirements or it was a different location than after said label.

> Not all countries use asterisks or bold type to denote a required field. For example the “※” symbol is a popular in Japan to show that something is important or deserves attention. An asterisk works as well, but “※” is so particular to Japan that it could be used to build consumer confidence so customers know they are dealing with a local company (or at least one that researched the market).

[Source](https://www.cleverbridge.com/corporate/3-tips-to-localize-your-form-fields/)

haverchuck replied:
@seang96 Ok, that makes sense.  Thanks.

haverchuck commented on packages/aws-amplify-angular/src/components/authenticator/sign-up-component/sign-up.component.core.ts:
```diff
@@ -24,45 +24,43 @@ const template = `
           <label class="amplify-input-label">
-            {{field.label}} 
-            <span *ngIf="field.required">*</span>
+            {{ this.amplifyService.i18n().get(field.label) }}
```
@seang96 Please reenable `<span *ngIf="field.required">*</span>`.  This will automatically insert an asterix if the field is required.

haverchuck commented on packages/aws-amplify-angular/src/components/authenticator/sign-up-component/sign-up.component.core.ts:
```diff
@@ -24,45 +24,43 @@ const template = `
           <label class="amplify-input-label">
-            {{field.label}} 
-            <span *ngIf="field.required">*</span>
+            {{ this.amplifyService.i18n().get(field.label) }}
```
@seang96 Please see above.

haverchuck commented on packages/aws-amplify-angular/src/components/authenticator/sign-up-component/sign-up.component.ionic.ts:
```diff
@@ -39,70 +38,67 @@ const template = `
                   type={{field.type}}
                   class="amplify-form-input-phone-ionic"
-                  placeholder={{field.label}}
+                  [placeholder]="this.amplifyService.i18n().get(field.placeholder)"
```
@seang96 Good catch, thank you!

haverchuck commented on packages/pubsub/package.json:
```diff
@@ -11,7 +11,7 @@
     "test": "tslint 'src/**/*.ts' && jest --coverage",
     "build-with-test": "npm run clean && npm test && tsc && webpack",
-    "build": "npm run clean && tsc && cp -r src/vendor lib/vendor && webpack",
+    "build": "npm run clean && tsc && cpx src/vendor/** lib/vendor && webpack",
```
Please remove these changes to PubSub package.json. If needed, please submit them in a different PR.

haverchuck commented on packages/pubsub/package.json:
```diff
@@ -35,6 +35,7 @@
     "babel-preset-react": "^6.24.1",
     "babel-preset-stage-2": "^6.24.1",
     "compression-webpack-plugin": "^1.1.3",
+    "cpx": "^1.5.0",
```
@seang96 Please see above.

seang96 commented:
@haverchuck I removed the package.json modifications and put them in another PR as requested.

I also just added a commit that fixed the phone number section in the sign up form for Ionic that would make it scroll on small screens. I do not think this was an issue caused by my changes, but I updated just to make sure.

haverchuck commented:
@seang96 I thought about your point that the asterix does not universally denote a required field, and I think it makes sense.  However, I think we should retain it for the time being since a) our other framework packages contain this, and b) it would be considered a breaking change (albeit very minor breakage).

Removing the asterix would make sense as part of a major version change.

seang96 commented:
> @seang96 I thought about your point that the asterix does not universally denote a required field, and I think it makes sense. However, I think we should retain it for the time being since a) our other framework packages contain this, and b) it would be considered a breaking change (albeit very minor breakage).
> 
> Removing the asterix would make sense as part of a major version change, across framework packages. Your other changes are great and I would love to merge them ASAP.

I can agree with that, I did a new commit to revert back to the old way for those fields.
